test/server/server_test.cc

```diff
@@ -334,6 +334,60 @@ INSTANTIATE_TEST_SUITE_P(IpVersions, ServerInstanceImplTest,
                          testing::ValuesIn(TestEnvironment::getIpVersionsForTest()),
                          TestUtility::ipTestParamsToString);
 
+// The finalize() of Http::CustomInlineHeaderRegistry will be called When InstanceImpl is
+// instantiated. For this reason, the test case for inline headers must precede all other test
+// cases.
+
+// Test the case where inline headers contain names that cannot be registered as inline headers.
+TEST_P(ServerInstanceImplTest, WithErrorCustomInlineHeaders) {
+  EXPECT_THROW_WITH_MESSAGE(
+      initialize("test/server/test_data/server/bootstrap_inline_headers_error.yaml"),
+      EnvoyException, "Header set-cookie cannot be registered as an inline header.");
+}
+
+TEST_P(ServerInstanceImplTest, WithDeathCustomInlineHeaders) {
+#if !defined(NDEBUG)
+  // The finalize() of Http::CustomInlineHeaderRegistry will be called after the first successful
+  // instantiation of InstanceImpl. If InstanceImpl is instantiated again and the inline header is
+  // registered again, the assertion will be triggered.
+  EXPECT_DEATH(
+      {
+        initialize("test/server/test_data/server/bootstrap_inline_headers.yaml");
+        initialize("test/server/test_data/server/bootstrap_inline_headers.yaml");
+      },
+      "");
+#endif // !defined(NDEBUG)
+}
+
+// Test whether the custom inline headers can be registered correctly.
+TEST_P(ServerInstanceImplTest, WithCustomInlineHeaders) {
+  static bool is_registered = false;
+
+  if (!is_registered) {
+    // Avoid repeated registration of custom inline headers in the current process after the
+    // finalize() of Http::CustomInlineHeaderRegistry has already been called.
+    EXPECT_NO_THROW(initialize("test/server/test_data/server/bootstrap_inline_headers.yaml"));
+    is_registered = true;
+  }
+
+  EXPECT_TRUE(
```

Comment:
how about an integration test (protocol integration test?) that the headers / trailers end up elided?

Reply:
get it. I think we can verify this problem with a few simple lines of tests in existing test cases.

Reply:
Where did we land on this? Don't see an integration test